spec/unit/ace/file_mutex_spec.rb

```diff
@@ -0,0 +1,188 @@
+
+  before do
+    File.delete(lock_file) if File.exist?(lock_file)
+  end
+
+  describe '#with_read_lock' do
+    let(:p1_message) { 'process 1 written' }
+    let(:p2_message) { 'process 2 written' }
+
+    it 'creates the lock file' do
+      foo = mutex.with_read_lock do
+        lock_content
+      end
+
+      expect(File).to exist(lock_file)
+      expect(foo).to eq(lock_content)
+    end
+
+    it 'allows multiple non-blocking calls' do
+      p1_in_read, _p1_in_write = IO.pipe
+      p1_out_read, p1_out_write = IO.pipe
+      p2_in_read, _p2_in_write = IO.pipe
+      p2_out_read, p2_out_write = IO.pipe
+
+      Timeout.timeout(2) do
+        fork do
+          mutex.with_read_lock do
+            p1_out_write.puts p1_message
+
+            # start the blocking read
+            puts p1_in_read.read
+
+            # tidy up
+            p1_out_write.close
+          end
+        end
+        fork do
+          mutex.with_read_lock do
+            p2_out_write.puts p2_message
+
+            # start the blocking read
+            puts p2_in_read.read
+
+            # tidy up
+            p2_out_write.close
+          end
+        end
+        # close the blocking reads
+        p1_in_read.close
+        p2_in_read.close
+
+        # now check that each fork performed the operations at the same time
+        expect(p1_out_read.gets).to match(/#{p1_message}/)
+        expect(p2_out_read.gets).to match(/#{p2_message}/)
+      end
+
+      # tidy up
+      p1_out_read.close
+      p2_out_read.close
+    end
+  end
+
+  describe '#with_write_lock' do
+    it 'creates the lock file' do
+      foo = mutex.with_write_lock do
+        lock_content
+      end
+
+      expect(File).to exist(lock_file)
+      expect(foo).to eq(lock_content)
+    end
+
+    context 'when a write lock is held' do
+      let(:p1_message) { 'process 1 written' }
+      let(:p2_message) { 'process 2 should not write' }
+
+      p1_in_read, _p1_in_write = IO.pipe
+      p1_out_read, p1_out_write = IO.pipe
+      p2_in_read, _p2_in_write = IO.pipe
+      p2_out_read, p2_out_write = IO.pipe
+      exception_thrown = false
+
+      it 'second lock is blocked' do
+        # this should perform its operation
+        fork do
+          mutex.with_write_lock do
+            p1_out_write.puts p1_message
+
+            # start the blocking read
+            puts p1_in_read.read
+
+            # tidy up
+            p1_out_write.close
+          end
+        end
+
+        begin
+          # this should block
+          Timeout.timeout(1) do
+            fork do
+              mutex.with_write_lock do
+                p2_out_write.puts p2_message
+
+                # start the blocking read
+                puts p2_in_read.read
+
+                # tidy up
+                p2_out_write.close
+              end
+            end
+
+            # these will force the mutex
+            p1_out_read.gets
+            p2_out_read.gets
+          end
+        rescue StandardError
```

Comment:
Please be specific in what exceptions this should catch. You can use
```ruby
expect do
   # ...
end.to raise_error Timeout::Error
```
to avoid the `be_truthy` below.